### ISSUE TITLE
Point-Wise Logical Analysis

### ISSUE THREAD

**Issue by lightsighter** [StanfordLegion/legion#1458]
Reporting on behalf of @magnatelee as this is a feature request from the Legate/cuNumeric team.

Currently, Legion's logical dependence analysis is sound but imprecise when it comes to analyzing dependences between index space tasks/operations. For example, if two index space tasks have interfering projection region requirements, then all the points in the first index space task must map before any of the points in the second index space task can be mapped. This is unfortunate in cases where the dependences between point tasks are "simple" (e.g. data parallel across the points), and prevents mappers from mapping points from later index launches as soon as their dependences from earlier index launches are satisfied. For example, in the following program all three index spaces projection from the same partition `P` with the identity projection functor:

```
Index Space Task A: [0-3] P[identity]
Index Space Task B: [0-3] P[identity]
Index Space Task C: [0-3] P[identity]
```

It should be possible to map point tasks 0 for `A` and then `B` and then `C` before mapping any other point tasks. 

The goal of this feature would be to make the logical dependence analysis precise enough to allow point-wise mapping dependences between index space tasks/operations whenever their dependence pattern was "simple". The proposed definition of "simple" would be when the dependence analysis can be done with a name-based analysis instead of a content-based analysis. Any failure of name-based analysis would immediately fall back to the current implementation and enforce that all the points between the first index space launch would be mapped before any in the second index space launch. 

Complication: this would need to work across shards in control replicated contexts as well.

Dependence: the implementation of this feature will depend on some machinery currently in the `shardrefine` branch.

Side benefit: the same analysis that we would have to do for this feature, would probably allow us to elide the requirement that mappers tell us where to do collective analyses for particular region requirements. We'll be able to identify that automatically from the analysis required here. (@ipdemes for visibility)

**Comment by rohany** [comment]
Can you elaborate the case where it's beneficial to be able to map these point tasks ahead of the full index launch?

**Comment by lightsighter** [comment]
> Does a pointwise logical analysis mean that it is still O(1) in partition size?

If the two index launches are using the same pure functional projection function, projecting from the same upper bound, and the launch domain of the second index launch is a (non-strict) subset of the launch domain of the first index launch then the analysis will be O(1) in the number of logical regions used in the subtree. For all other cases it will be O(N) in the number of logical regions used in the subtree. For pure functional projection functors though the analysis will be memoized in a cache to reduce analysis overheads. We actually already have a variation of this analysis in the `shardrefine` branch to figure out where to insert merge close op fences across shards for when dependences span multiple shards.

> Will you only support the identity projection functor or will user-defined projection functors be supported, and if so with what constraints?

We will support all projection functors, but pure functional projection functors (as defined by #1410) will support memoization to reduce overheads. Non-functional projection functors will incur the full analysis all the time.

> Can you elaborate the case where it's beneficial to be able to map these point tasks ahead of the full index launch?

The primary use case is for users that want to do strip-mined execution of points across index space launches to reduce aggregate memory footprints of all the point tasks running in parallel.

**Comment by elliottslaughter** [comment]
> For all other cases it will be O(N) in the number of logical regions used in the subtree.

Is this really `O(N/P)` where `P` is the number of shards? Or truly `O(N)` (i.e., all shards need to do all the work)?

I am a bit concerned about adding an expensive analysis here. I understand it will be cached in most cases we care about, but I still think that if we build this, we should have someone do a performance evaluation before we merge it into any branches we care about.

**Comment by lightsighter** [comment]
> Is this really O(N/P) where P is the number of shards? Or truly O(N) (i.e., all shards need to do all the work)?

With the new machinery in `shardrefine` it will be O(N/P) but will require a synchronization across the shards to perform a boolean all-reduce to make sure they all agree on the answer in the case that we actually do compute it and can't just look up a memoized result (this is a synchronization across the shards in the logical analysis, so a bit unfortunate, but not too terrible if memoized). If you're curious I can share details about the representation of the projection and how we do the analysis efficiently.

**Comment by elliottslaughter** [comment]
Can you outline the cache logic? I assume it's something like same projection ID, projection args (if applicable), and index space domain?

> require a synchronization across the shards to perform a boolean all-reduce to make sure they all agree on the answer in the case that we actually do compute it and can't just look up a memoized result

This makes it sound like you're doing communication even in the cache hit case, which would be unfortunate.

**Comment by lightsighter** [comment]
> Can you outline the cache logic? I assume it's something like same projection ID, projection args (if applicable), and index space domain?

You can see the function that does that "matching" for entries in the cache here:
https://gitlab.com/StanfordLegion/legion/-/blob/shardrefine/runtime/legion/legion_analysis.cc#L3995-4014
If the function returns `true` then you've hit in the cache. If it returns `false` then you miss and we go on to the next entry.

> This makes it sound like you're doing communication even in the cache hit case, which would be unfortunate.

No, we only do communication when first computing the [summary properties](https://gitlab.com/StanfordLegion/legion/-/blob/shardrefine/runtime/legion/legion_analysis.h#L1291-1308) of the projection. The cache is deterministic across the shards of control replication so all the shards will hit and return the same precomputed value or they will all miss and do their local analysis and then exchange results to make sure they all agree on the results.

You'll notice that one of those summary properties is whether it supports "self name-based analysis" meaning if you see back-to-back index launches of the same projection summary you'll know immediately whether you do point-to-point. In the case of comparing two different projection summaries, there will be another cache for that to record whether two different projection summaries support name-based analysis between them or not (also deterministically replicated across the shards).

**Comment by elliottslaughter** [comment]
I think there's a bug in your code. I see:

```c++
if ((arglen > 0) && (memcmp(args, projection_args, arglen) == 0))
```

But `memcmp` returns 0 when the values are equal: https://cplusplus.com/reference/cstring/memcmp/

**Comment by elliottslaughter** [comment]
Your cache appears to check exact equality in the projection and sharding spaces, but I think we could support (potentially non-strict) subsets?

**Comment by lightsighter** [comment]
> I think there's a bug in your code

Yes, you can tell I haven't run this code yet. It still doesn't link.

> Your cache appears to check exact equality in the projection and sharding spaces, but I think we could support (potentially non-strict) subsets?

We could potentially relax the equality check on the domain of the index space launch to match on any new index spaces which subsets of the original domain and still be sound, but then we would lose precision in determining if a name-based analysis is permitted or not, which I don't think will be allowed in this new feature. If you can do the analysis with a name-based analysis then it has to work, so we must be both sound and precise, therefore no under approximations of the launch space domain are permitted in the cache.

As for the sharding space domain, I don't think it's ever sound to do anything other exact matching because we don't know how the sharding functions are going to use it.

**Comment by lightsighter** [comment]
We're close to finishing this particular issue. Pointwise dependence analysis has been in the master branch since:
https://gitlab.com/StanfordLegion/legion/-/merge_requests/1619
Support for pointwise dependence on output regions is nearing completion:
https://gitlab.com/StanfordLegion/legion/-/merge_requests/1797
The one remaining thing to decide for this issue is how to expose pointwise analysis. Currently it is guarded behind a command line flag that enables pointwise analysis for all tasks if passed. It's unclear if we want to just turn pointwise analysis on all the time or if the extra overhead it incurs is too high for that. Initial experiments suggest the overhead is not unreasonably large, but it is measurable.
